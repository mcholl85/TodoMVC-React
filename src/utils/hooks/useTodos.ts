import { useMutation, useQuery, useQueryClient, UseQueryResult } from '@tanstack/react-query'
import { createTodo, deleteTodo, deleteTodos, fetchTodos, updatePartiallyTodo } from '../api/todos'

export interface TodoType {
  id: string
  title: string
  completed: boolean
}

export function useTodos(filter: string) {
  const queryClient = useQueryClient()
  const { data: todos }: UseQueryResult<TodoType[]> = useQuery({
    queryKey: ['todos'],
    queryFn: fetchTodos,
  })

  const { mutate: saveTodo } = useMutation({
    mutationFn: createTodo,
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['todos'] })
    },
  })

  const displayedTodos =
    todos?.filter((todo) => {
      switch (filter) {
        case 'active':
          return !todo.completed
        case 'completed':
          return todo.completed
        default:
          return true
      }
    }) || []

  const allTodosCompleted = displayedTodos?.every((todo) => todo.completed)

  const { mutate: removeTodo } = useMutation({
    mutationFn: deleteTodo,
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['todos'] })
    },
  })

  const { mutate: setTitle } = useMutation({
    mutationFn: updatePartiallyTodo,
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['todos'] })
    },
  })

  const { mutate: setCompleted } = useMutation({
    mutationFn: updatePartiallyTodo,
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['todos'] })
    },
  })

  const { mutate: clearCompleted } = useMutation({
    mutationFn: () => deleteTodos(true),
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['todos'] })
    },
  })

  const changeAllCompletedTodo = (bool: boolean) => {
    todos?.forEach((todo) => {
      if (todo.completed === bool) setCompleted({ id: todo.id, completed: bool })
    })
  }

  return {
    todos: displayedTodos,
    count: {
      all: todos?.length || 0,
      completed: todos?.filter((todo) => todo.completed).length || 0,
      active: todos?.filter((todo) => !todo.completed).length || 0,
    },
    allTodosCompleted,
    saveTodo,
    removeTodo,
    setTitle,
    setCompleted,
    clearCompleted,
    changeAllCompletedTodo,
  }
}
