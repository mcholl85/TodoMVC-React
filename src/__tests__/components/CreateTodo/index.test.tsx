import { render, renderHook, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import CreateTodo from '../../../components/CreateTodo'
import { useTodos } from '../../../utils/hooks/useTodos'

describe('Component CreateTodo', () => {
  test('Matches DOM Snapshot', () => {
    const { asFragment } = render(<CreateTodo saveTodo={() => []} />)

    expect(asFragment()).toMatchSnapshot()
  })
  test('change value of new todo input and create a todo after press enter', () => {
    const mockSaveTodo = jest.fn()
    render(<CreateTodo saveTodo={mockSaveTodo} />)

    const inputEl: HTMLInputElement = screen.getByRole('textbox')

    userEvent.type(inputEl, 'test')
    expect(inputEl).toHaveValue('test')

    userEvent.type(inputEl, '{enter}')
    expect(mockSaveTodo).toBeCalledWith('test')
    expect(inputEl).toHaveValue('')
  })
  test('do not create a todo given a empty input', () => {
    const { result } = renderHook(() => useTodos('all'))

    render(<CreateTodo saveTodo={result.current.saveTodo} />)

    const inputEl: HTMLInputElement = screen.getByRole('textbox')

    userEvent.type(inputEl, '{enter}')

    expect(result.current.todos).toEqual([])
  })
})
