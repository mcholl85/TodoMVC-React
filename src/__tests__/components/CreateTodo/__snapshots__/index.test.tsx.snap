// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component CreateTodo Matches DOM Snapshot 1`] = `
<DocumentFragment>
  <input
    class="new-todo"
    placeholder="What needs to be done?"
    value=""
  />
</DocumentFragment>
`;
